### PR TITLE
KuCoin auth method

### PR DESCRIPTION
**Pull request by peedrr**
### Updates KuCoin authentication to conform to the new signing method

- [X] - Tested - used NBBO and confirmed updates being received
- [X] - Changelog updated
- [ ] - Tests run and pass - I couldn't see any tests relating to KuCoin. Please advise if needed
- [X] - Flake8 run and all errors/warnings resolved
- [ ] - Contributors file updated (optional)

### CODE REVIEW

File: cryptofeed/auth/kucoin.py [hunk 1]
```diff
@@ -13,16 +13,19 @@
...
     header = {
-        "KC-API-SIGN": signature.decode(),
-        "KC-API-TIMESTAMP": str(now),
         "KC-API-KEY": key_id,
-        "KC-API-PASSPHRASE": key_passphrase,
-        "KC-API-KEY-VERSION": "3"
+        "KC-API-SIGN": signature.decode(),
+        "KC-API-TIMESTAMP": now,
+        "KC-API-PASSPHRASE": passphrase.decode(),
+        "KC-API-KEY-VERSION": "2"
```
Comment (bmoscon) [inline cryptofeed/auth/kucoin.py:29]: I'm un-clear why this header needs to change? I tested the kucoin auth with the current version in the code and didn't have any issues with it

**Review by bmoscon — CHANGES_REQUESTED** [review]

File: CHANGES.md [hunk 2]
```diff
@@ -1,5 +1,8 @@
 ## Changelog
 
+### 1.9.3
```
[the remark this answers is not in the document]
Comment (peedrr) [reply CHANGES.md:3]: I took a punt on this - no problem, will add it to previous version

File: cryptofeed/auth/kucoin.py (hunk 1 above)
Comment (peedrr) [reply cryptofeed/auth/kucoin.py:29]: There is no v3 of the API _**Key**_. (The API itself spans v1-3, depending on the method)

They have [only just deprecated v1](https://www.kucoin.com/news/en-announcement-regarding-the-postponement-of-kucoin-api-key-2-upgrade) so I took this bit of code to be an error.

From [the docs](https://docs.kucoin.top/#authentication): 

"KC-API-KEY-VERSION You can check the version of API key on the page of API Management 
_(All my new ones are automatically v2)_
...
For the KC-API-PASSPHRASE of the header:

For API key-V1.0, please pass requests in plaintext.
For API key-V2.0, please Specify KC-API-KEY-VERSION as 2 "
_(No mention of v3 anywhere)_

**Review by bmoscon — APPROVED** [review]

File: cryptofeed/auth/kucoin.py (hunk 1 above)
Comment (tudorelu) [reply cryptofeed/auth/kucoin.py:29]: Did you try to run with your new config? I know their docs state that the passphrase needs to be a signed key, but that didn't work for me. I wrote that function specifically for the order book endpoint, which if you check the documentation will see that v2 was deprecated, and v3 is left. It was working well when I left it.

https://docs.kucoin.com/#get-full-order-book-aggregated

**Comment by tudorelu** [comment]
Somehow I doubt that this will work. The function `generate_token` was written specifically for the order_books endpoint, which has been recently updated to v3 (v2 has been deprecated for that endpoint) https://docs.kucoin.com/#get-full-order-book-aggregated. As far as cryptofeed supports, only order books need authentication for now.

Also, even though their docs say that the passphrase needs to be the encrypted secret key, in practice that is not the case and the way it was before your changes was working (at least for the order books endpoint, the passphrase is simply your api key's password). 

You can test your changes by trying to run examples/demo_kucoin_authenticated.py, but make sure you have passed correct api_key, secret_key and passphrase in the config.

**Comment by peedrr** [comment]
Thanks @tudorelu for the pointer to the demo code. I'll test and report back

**Comment by peedrr** [comment]
**TL;DR**
I've tested both demo_nbbo.py and demo_kucoin_authenticated.py and I can confirm that the current master is in a good, working state.

**Steps taken to test:**
I started by rolling back to [https://github.com/bmoscon/cryptofeed/commit/9dd1918ef2a6cc5f384aecb07f28d0e683c24dc2](9dd191)
```git checkout 9dd1918ef2a6cc5f384aecb07f28d0e683c24dc2 ```
and, as you rightly said, the auth worked fine with examples/demo_kucoin_authenticated.py

I then pulled in @bmoscon's [update to NBBO](https://github.com/bmoscon/cryptofeed/commit/3f608481512a9b3ef61b4ec4732b9e787b07c0f0) which allows passing of the config through `add_nbbo`
```git cherry-pick 3f608481512a9b3ef61b4ec4732b9e787b07c0f0```

When I run demo_nbbo.py (with valid config), I get a 401 response:
```Traceback (most recent call last):
  File "/Users/peteh/Documents/programming/trading/github-repos/cryptofeed/cryptofeed/connection_handler.py", line 64, in _create_connection
    await self._handler(connection, self.handler)
  File "/Users/peteh/Documents/programming/trading/github-repos/cryptofeed/cryptofeed/connection_handler.py", line 113, in _handler
    await handler(message, connection, self.conn.last_message)
  File "/Users/peteh/Documents/programming/trading/github-repos/cryptofeed/cryptofeed/exchange/kucoin.py", line 225, in message_handler
    await self._l2_book(msg, symbol, timestamp)
  File "/Users/peteh/Documents/programming/trading/github-repos/cryptofeed/cryptofeed/exchange/kucoin.py", line 185, in _l2_book
    await self._snapshot(symbol)
  File "/Users/peteh/Documents/programming/trading/github-repos/cryptofeed/cryptofeed/exchange/kucoin.py", line 150, in _snapshot
    data = await self.http_conn.read(url, header=headers)
  File "/Users/peteh/Documents/programming/trading/github-repos/cryptofeed/cryptofeed/connection.py", line 129, in read
    response.raise_for_status()
  File "/Users/peteh/.local/share/virtualenvs/cryptofeed-XtJ7R2bh/lib/python3.9/site-packages/aiohttp/client_reqrep.py", line 1000, in raise_for_status
    raise ClientResponseError(
aiohttp.client_exceptions.ClientResponseError: 401, message='Unauthorized', url=URL('https://api.kucoin.com/api/v3/market/orderbook/level2?symbol=BTC-USDT')```

I then pulled in the [reworking of the signing method](https://github.com/bmoscon/cryptofeed/commit/efeb4d703428cfc26f05e418131510aaea718973)...

```git cherry-pick efeb4d703428cfc26f05e418131510aaea718973```
and the NBBO demo is again working as expected:
```Pair: BTC-USDT Bid Price: 35820.66 Bid Size: 0.094769 Bid Feed: COINBASE Ask Price: 35822.71 Ask Size: 0.205000 Ask Feed: COINBASE
Pair: BTC-USDT Bid Price: 35820.67 Bid Size: 0.006701 Bid Feed: COINBASE Ask Price: 35822.71 Ask Size: 0.205000 Ask Feed: COINBASE
Pair: BTC-USDT Bid Price: 35820.66 Bid Size: 0.094769 Bid Feed: COINBASE Ask Price: 35822.71 Ask Size: 0.205000 Ask Feed: COINBASE
Pair: BTC-USDT Bid Price: 35820.67 Bid Size: 0.006701 Bid Feed: COINBASE Ask Price: 35822.71 Ask Size: 0.205000 Ask Feed: COINBASE
Pair: BTC-USDT Bid Price: 35820.67 Bid Size: 0.006701 Bid Feed: COINBASE Ask Price: 35817.60 Ask Size: 0.099642 Ask Feed: KUCOIN
Pair: BTC-USDT Bid Price: 35820.67 Bid Size: 0.006701 Bid Feed: COINBASE Ask Price: 35817.60 Ask Size: 0.179642 Ask Feed: KUCOIN
Pair: BTC-USDT Bid Price: 35822.70 Bid Size: 0.120000 Bid Feed: COINBASE Ask Price: 35817.60 Ask Size: 0.179642 Ask Feed: KUCOIN
Pair: BTC-USDT Bid Price: 35822.70 Bid Size: 0.120000 Bid Feed: COINBASE Ask Price: 35817.60 Ask Size: 0.174642 Ask Feed: KUCOIN```

Finally, I switched back to origin/master, put valid keys in config.yaml and ran  demo_kucoin_authenticated.py again. Output is as expected:
```...

                  Decimal('2369.9'): Decimal('0.0003834'),
                  Decimal('2370'): Decimal('0.1674571'),
                  Decimal('2370.08'): Decimal('18.5074013'),
                  Decimal('2370.09'): Decimal('0.0009174'),
                  Decimal('2370.15'): Decimal('0.1'),
                  Decimal('2370.26'): Decimal('0.0004413'),
                  Decimal('2370.28'): Decimal('0.05'),
                  Decimal('2370.31'): Decimal('0.02')}},
 'feed': 'KUCOIN',
 'receipt_timestamp': 1623586489.448886,
 'symbol': 'ETH-USDT',
 'timestamp': 1623586489.448886}```
 
 I'm still exploring the library so I don't know exactly what calls demo_nbbo is making which require the new form of auth but, I'm quite sure, by using the new form of auth, we haven't broken anything.
 
 Of course, as original author of the KuCoin auth function, I would be grateful if you would perform the same or similar tests to confirm you are happy as well.
 
 Cheers.

**Comment by bmoscon** [comment]
the authentication code in cryptofeed is nearly identical to what kucoin lists as an example in their docs. this change works for me, so nothing seems to be broken and it matches the docs.

**Comment by tudorelu** [comment]
Great, can confirm that your changes work on my end as well @peedrr 

My confusion came from the fact that my changes also worked, but in my case the headers were:
```py
    header = {
        "KC-API-SIGN": signature.decode(),
        "KC-API-TIMESTAMP": str(now),
        "KC-API-KEY": key_id,
        **"KC-API-PASSPHRASE": key_passphrase,**
        **"KC-API-KEY-VERSION": "3"**
```

Whereas yours looked like:
```py
    header = {
        "KC-API-KEY": key_id,
        "KC-API-SIGN": signature.decode(),
        "KC-API-TIMESTAMP": now,
        **"KC-API-PASSPHRASE": passphrase.decode(),**
        **"KC-API-KEY-VERSION": "2"**
    }
```

But turns out both work!

**Comment by peedrr** [comment]
Cool. Thanks for confirming
